# Data Sources
general gene expression dbs: 
- https://www.ncbi.nlm.nih.gov/geo/
- https://www.ebi.ac.uk/biostudies/arrayexpress
- https://www.ensembl.org/
- http://m3d.mssm.edu/
- http://regulondb.ccg.unam.mx/
## Beta Estimation
### time series
- https://zenodo.org/records/8114649
  - ai training data; not sure the exact formating (article: https://www.nature.com/articles/s41467-024-46361-1#Sec2)
- 

### tf knockout

### static measurement
- https://github.com/makselab/geneCoexpressionFibration/tree/main/ExpressionData
  - contains two files with tons of static measurements in different conditions
- https://genexpdb.okstate.edu/
  - vast amounts of single measurement snapshots of each gene's expression; specific to e coli
- 

## Binding Site Inference
- currently have ~2500 transcription factor - binding motif pairs; from memedb


## just brainstorming
- if extracting constant betas is difficult, perhaps ml can be used to estimate betas combined with binding location
